contracts/root/predicates/ERC20Predicate.sol

```diff
@@ -0,0 +1,283 @@
+    require(
+      bytes32(inputItems[0].toUint()) == WITHDRAW_EVENT_SIG,
+      "Not a withdraw event signature"
+    );
+    address rootToken = address(RLPReader.toUint(inputItems[1]));
+    require(
+      msg.sender == address(inputItems[2].toUint()), // from
+      "Withdrawer and burn exit tx do not match"
+    );
+    uint256 exitAmount = BytesLib.toUint(logData, 0); // amountOrTokenId
+    withdrawManager.addExitToQueue(msg.sender, childToken, rootToken, exitAmount, bytes32(0x0), true /* burnt */, age);
+  }
+
+  function startExit(bytes calldata data, bytes calldata exitTx)
+    external
+  {
+    RLPReader.RLPItem[] memory referenceTxData = data.toRlpItem().toList();
+    uint256 age = withdrawManager.verifyInclusion(data, 0 /* offset */, false /* verifyTxInclusion */);
+    // validate exitTx - This may be an in-flight tx, so inclusion will not be checked
+    (uint256 exitAmount, address childToken, address participant, bytes32 txHash, bool burnt) = processExitTx(exitTx);
+
+    // process the receipt of the referenced tx
+    address rootToken;
+    uint256 closingBalance;
+    (rootToken, closingBalance, age) = processReferenceTx(
+      referenceTxData[6].toBytes(), // receipt
+      referenceTxData[9].toUint(), // logIndex
+      participant, // to verify that the balance of the signer of the exit tx is being referenced here
+      childToken,
+      age,
+      false /* isChallenge */);
+
+    // The closing balance of the exitTx should be <= the referenced balance
+    require(
+      closingBalance >= exitAmount,
+      "Exiting with more tokens than referenced"
+    );
+    if (referenceTxData.length > 10) {
+      // It means the exitor sent along another input UTXO to the exit tx.
+      // This will be used to exit with the pre-existing balance on the chain along with the couterparty signed exit tx
+      uint256 age2 = withdrawManager.verifyInclusion(data, 10 /* offset */, false /* verifyTxInclusion */);
+      address _rootToken;
+      (_rootToken, closingBalance, age2) = processReferenceTx(
+        referenceTxData[16].toBytes(), // receipt
+        referenceTxData[19].toUint(), // logIndex
+        msg.sender, // participant
+        childToken,
+        age2,
+        false /* isChallenge */);
+      require(rootToken == _rootToken, "root tokens in the referenced txs do not match"); // might not require this check
+      uint256 priority = Math.max(age, age2);
+      withdrawManager.addExitToQueue(msg.sender, childToken, rootToken, exitAmount + closingBalance, txHash, burnt, priority);
+      withdrawManager.addInput(priority, age, participant);
+      withdrawManager.addInput(priority, age2, msg.sender);
+    } else {
+      withdrawManager.addExitToQueue(msg.sender, childToken, rootToken, exitAmount, txHash, burnt, age);
+      withdrawManager.addInput(age, age, participant);
+    }
+  }
+
+  function verifyDeprecation(bytes calldata exit, bytes calldata inputUtxo, bytes calldata challengeData)
```

Comment:
Can we have gas estimation for ERC20/721?

Reply:
https://github.com/maticnetwork/contracts/pull/78#issuecomment-498642475